### PR TITLE
[BACKEND-GUILD] Add and auto-set new relic layer

### PR DESCRIPTION
**Pull request by wmcqueenzava**
## Overview
Instrument SAM lambda with New Relic

## Problem
Add the appropriate new relic layer to the lambda, and wrap the lambda with the new relic wrapper

## Change log
- Add and auto-set new relic layer

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue).
- [X] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). This requires a major version bump.
- [ ] This change requires a documentation update.

## How can this be tested?
Build, lint, test and deploy

## Checklist

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: template.yaml [hunk 1]
```diff
@@ -24,11 +24,11 @@ Parameters:
 Mappings:
   StagesMap:
     sandbox:
-      VariableName: Value
+      NewrelicExtensionLayer: NEW_RELIC_LAYER_NODE16
```
Comment (chibuzorugwunna) [inline template.yaml:27]: Interesting stuff - does SAM automatically pickup the `NEW_RELIC_LATEST_LAYER_NODE16` the current session?

Comment (wmcqueenzava) [reply template.yaml:27]: SAM doesn't, no.  I use the string to do a find and replace in the `deploy.sh` script.

Comment (enGMzizo) [reply template.yaml:27]: @chibuzorugwunna it gets replaced by the latest version of newrelic here 
```
sed -i'.layer.bak' -e "s/NEW_RELIC_LAYER_NODE16/$NEW_RELIC_LATEST_LAYER_NODE16/g" build/templates/$1-$4-out.yaml
```

File: scripts/deploy.sh [hunk 2]
```diff
@@ -3,5 +3,15 @@
 # Package binaries and upload them to S3 bucket
 sam package --template-file .aws-sam/build/$4.yaml --s3-bucket $2 --output-template-file build/templates/$1-$4-out.yaml --profile $1
 
+# Wrap lambdas in the newrelic wrapper handler
+sed -i'.handler.bak' -e 's/      Handler: app.lambdaHandler/      Handler: newrelic-lambda-wrapper.handler/g' build/templates/$1-$4-out.yaml
+
+# Set NewRelic Layer Version
+NEW_RELIC_LAYER_API_URL="https://eu-west-1.layers.newrelic-external.com/get-layers"
```
Comment (wmcqueenzava) [inline scripts/deploy.sh:10]: We should probably use `$AWS_REGION` here to select the appropriate region

File: scripts/deploy.sh [hunk 3]
```diff
@@ -3,5 +3,15 @@
...
+NEW_RELIC_LATEST_LAYER_NODE16=$(curl -s ${NEW_RELIC_LAYER_API_URL}?CompatibleRuntime=nodejs16.x \
+  | jq -r '.Layers[] | select(.LayerArn == "arn:aws:lambda:eu-west-1:451483290750:layer:NewRelicNodeJS16X") | .LatestMatchingVersion.LayerVersionArn'
+)
+sed -i'.layer.bak' -e "s/NEW_RELIC_LAYER_NODE16/$NEW_RELIC_LATEST_LAYER_NODE16/g" build/templates/$1-$4-out.yaml
```
Comment (enGMzizo) [inline scripts/deploy.sh:14]: Looks good also I suggest we pass it as a parameter when we deploy the stack like the following 
```
sam deploy --template-file build/templates/$1-$4-out.yaml --s3-bucket $2 --stack-name $3 --capabilities CAPABILITY_NAMED_IAM --profile $1 --no-fail-on-empty-changeset --parameter-overrides Stage=$1 Username=$5 NewrelicExtensionLayer=$NEW_RELIC_LATEST_LAYER_NODE16
```
and in the `template.yaml` 
```
Parameters:
  NewrelicExtensionLayer:
    Type: String
    Description: NewRelic extension layer
```
Then reference it as a parameter 
```
Layers: 
 !Ref: UserName
```

File: template.yaml (hunk 1 above)
Comment (chibuzorugwunna) [reply template.yaml]: Ah I didn't know what that did tbh. Thanks @wmcqueenzava @enGMzizo 😄

File: template.yaml [hunk 4]
```diff
@@ -14,6 +14,10 @@ Globals:
         AccountId: !Ref AWS::AccountId
         Region: "eu-west-1"
         NEWRELIC_HANDLER: newrelic-lambda-wrapper.handler
```
Comment (enGMzizo) [inline template.yaml:16]: Why do we need this `NEWRELIC_HANDLER` ? variable

File: template.yaml [hunk 5]
```diff
@@ -11,24 +11,37 @@ Globals:
     Timeout: 5
     Environment:
       Variables:
+        AccountId: !Ref AWS::AccountId
         Region: "eu-west-1"
+        NEWRELIC_HANDLER: newrelic-lambda-wrapper.handler
```
Comment (enGMzizo) [inline template.yaml:16]: I can confirm, `NEWRELIC_HANDLER` is not used

**Pull request closed without merging**